memory/PoolAllocator.h

```diff
@@ -39,8 +39,27 @@ class PoolAllocator
     }
 
 public:
+    class Deleter
+    {
+    public:
+        Deleter() : _allocator(nullptr) {}
+        Deleter(PoolAllocator<ELEMENT_SIZE>* allocator) : _allocator(allocator) {}
+
+        template <typename T>
+        void operator()(T* r)
+        {
+            if (_allocator)
```

Comment:
This should be only `nullptr` if the unique_ptr was default constructed. Then it only be possible to code hit here with a `nullptr` on `_allocator ` is when we use the `reset` method. Then I would suggest to make a validation in debug (like `assert`) just to ensure in debug we don't mess up